provider/oci/environ.go

```diff
@@ -57,30 +59,116 @@ func (e *Environ) ecfg() *environConfig {
 	return e.ecfgObj
 }
 
+func (e *Environ) allInstances(tags map[string]string) ([]*ociInstance, error) {
+	compartment := e.ecfg().compartmentID()
+	request := ociCore.ListInstancesRequest{
+		CompartmentId: compartment,
+	}
+	response, err := e.Compute.ListInstances(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ret := []*ociInstance{}
+	for _, val := range response.Items {
+		if val.LifecycleState == ociCore.InstanceLifecycleStateTerminated {
+			continue
+		}
+		missingTag := false
+		for i, j := range tags {
+			tagVal, ok := val.FreeformTags[i]
+			if !ok || tagVal != j {
+				missingTag = true
+				break
+			}
+		}
+		if missingTag {
+			// One of the tags was not found
+			continue
+		}
+		inst, err := newInstance(val, e)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ret = append(ret, inst)
+	}
+	return ret, nil
+}
+
+func (e *Environ) getOCIInstance(id instance.Id) (*ociInstance, error) {
+	instanceId := string(id)
+	request := ociCore.GetInstanceRequest{
+		InstanceId: &instanceId,
+	}
+
+	response, err := e.Compute.GetInstance(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return newInstance(response.Instance, e)
+}
+
 func (e *Environ) isNotFound(response *http.Response) bool {
 	if response.StatusCode == http.StatusNotFound {
 		return true
 	}
 	return false
 }
 
+// waitForResourceStatus will ping the resource until the fetch function returns true,
+// the timeout is reached, or an error occurs.
+func (o *Environ) waitForResourceStatus(
+	statusFunc func(resID *string) (status string, err error),
+	resId *string, desiredStatus string,
+	timeout time.Duration,
+) error {
+
+	var status string
+	var err error
+	timeoutTimer := o.clock.NewTimer(timeout)
+	defer timeoutTimer.Stop()
+
+	retryTimer := o.clock.NewTimer(0)
+	defer retryTimer.Stop()
+
+	for {
+		select {
+		case <-retryTimer.Chan():
+			status, err = statusFunc(resId)
+			if err != nil {
+				return err
+			}
+			if status == desiredStatus {
+				return nil
+			}
+			retryTimer.Reset(2 * time.Second)
```

Comment:
I think this got buried in the list of things to discuss. I am curious to consider what we want to do as a retry mechanism.

Reply:
It did. I could use the retry package, but it would not save much code. If anything, it will probably add some. The retry package seems to have been written to get around transient errors.

The point of this is not to retry in case of an error, but to poll for the status of a particular resource, given a function that fetches that resource status.It felt wrong to re-purpose a mechanism that is clearly centered around error handling (judging by the code and the function signature that gets passed in), to check statuses.

I would have to create a wrapper funtion that treats the fact that desired status != actual_status as an error, define a new error type that defines this as a transient error, and also define an ```IsFatalError()``` function that ignores it. So it just feels wrong in this scenario.

If the polling interval is too short, I can definitely change it. I don't think there is a need for a complicated back-off timer here, especially considering that in most places, there is a maximum timeout of 5 minutes, after which it will fail.